File: tools/pack_atlases.py
"""
pack_atlases.py — Pack PNG assets into texture atlases with Phaser JSON.

This script organizes all PNG sprites into multiple texture atlases:
  - atlas_characters.png/json — Mimi + enemies + NPCs
  - atlas_terrain.png/json    — floors, walls, decorations, landmarks
  - atlas_ui.png/json         — UI elements, hearts, items
  - atlas_backdrops.png/json  — battle backgrounds

Requires: pip install Pillow

Usage:
    python tools/pack_atlases.py
"""

import json
import re
from pathlib import Path
from PIL import Image

# Paths
REPO_ROOT = Path(__file__).parent.parent
ASSETS_DIR = REPO_ROOT / 'assets' / 'sprites'
ATLAS_OUTPUT_DIR = REPO_ROOT / 'assets' / 'atlases'
ASSET_CONFIG_PATH = REPO_ROOT / 'src' / 'config' / 'AssetConfig.js'

# Atlas configuration
MAX_ATLAS_SIZE = 2048  # 2048×2048 is widely supported
PADDING = 2            # Pixels between sprites to prevent bleeding


class Rect:
    """Simple rectangle for bin packing."""
    def __init__(self, x, y, width, height):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    
    def area(self):
        return self.width * self.height


class ShelfPacker:
    """
    Simple shelf-based bin packing algorithm.
    Packs rectangles onto horizontal shelves, creating new shelves as needed.
    """
    
    def __init__(self, max_width=MAX_ATLAS_SIZE, max_height=MAX_ATLAS_SIZE):
        self.max_width = max_width
        self.max_height = max_height
        self.shelves = []
        self.current_shelf = None
        self.used_height = 0
    
    def pack(self, width, height):
        """
        Try to pack a rectangle of given size.
        Returns (x, y) if successful, None if atlas is full.
        """
        # Add padding
        w = width + PADDING * 2
        h = height + PADDING * 2
        
        # Try to fit on current shelf
        if self.current_shelf:
            shelf_x, shelf_y, shelf_w, shelf_h = self.current_shelf
            next_x = shelf_x + shelf_w
            
            # Check if it fits on this shelf
            if next_x + w <= self.max_width and h <= shelf_h:
                # Fits on current shelf
                self.current_shelf = (next_x, shelf_y, shelf_w + w, shelf_h)
                return (next_x + PADDING, shelf_y + PADDING)
        
        # Need a new shelf
        new_shelf_y = self.used_height
        
        # Check if we have room for a new shelf
        if new_shelf_y + h > self.max_height:
            return None  # Atlas is full
        
        # Create new shelf
        self.current_shelf = (0, new_shelf_y, w, h)
        self.shelves.append(self.current_shelf)
        self.used_height = new_shelf_y + h
        
        return (PADDING, new_shelf_y + PADDING)
    
    def get_dimensions(self):
        """Return the actual used dimensions of the atlas."""
        if not self.shelves:
            return (0, 0)
        
        max_width = max(s[0] + s[2] for s in self.shelves)
        max_height = self.used_height
        
        # Round up to power of 2 for better GPU compatibility
        width = 1
        while width < max_width:
            width *= 2
        
        height = 1
        while height < max_height:
            height *= 2
        
        return (min(width, MAX_ATLAS_SIZE), min(height, MAX_ATLAS_SIZE))


def parse_asset_config():
    """Parse AssetConfig.js to categorize assets."""
    
    with open(ASSET_CONFIG_PATH, 'r', encoding='utf-8') as f:
        content = f.read()
    
    categories = {
        'characters': [],
        'terrain': [],
        'ui': [],
        'backdrops': []
    }
    
    # Map sections to categories
    section_map = {
        'SPRITE_DEFS': 'characters',
        'UI_DEFS': 'ui',
        'TERRAIN_DEFS': 'terrain',
        'BACKDROP_DEFS': 'backdrops'
    }
    
    for section, category in section_map.items():
        pattern = rf'export const {section} = \[(.*?)\];'
        match = re.search(pattern, content, re.DOTALL)
        
        if not match:
            continue
        
        array_content = match.group(1)
        
        # Parse each object
        obj_pattern = r'\{[^}]+\}'
        objects = re.findall(obj_pattern, array_content)
        
        for obj_str in objects:
            # Extract key and file
            key_match = re.search(r"key:\s*'([^']+)'", obj_str)
            file_match = re.search(r"file:\s*'([^']+)'", obj_str)
            
            if not key_match or not file_match:
                continue
            
            key = key_match.group(1)
            file = file_match.group(1)
            
            # Extract frames
            frames_match = re.search(r"frames:\s*\[([^\]]+)\]", obj_str)
            frames = []
            if frames_match:
                frames_str = frames_match.group(1)
                frames = re.findall(r"'([^']+)'", frames_str)
            
            categories[category].append({
                'key': key,
                'file': file,
                'frames': frames
            })
    
    return categories


def create_atlas(name, assets):
    """
    Create a texture atlas from a list of assets.
    Returns (atlas_image, atlas_json_data).
    """
    
    print(f"\n📦 Creating {name} atlas...")
    
    # Collect all images to pack
    images_to_pack = []
    
    for asset in assets:
        # Base image
        png_path = ASSETS_DIR / f"{asset['file']}.png"
        
        if not png_path.exists():
            print(f"  ⚠️  Missing: {png_path.name}")
            continue
        
        try:
            img = Image.open(png_path)
            images_to_pack.append({
                'key': asset['key'],
                'path': png_path,
                'image': img,
                'width': img.width,
                'height': img.height
            })
        except Exception as e:
            print(f"  ❌ Error loading {png_path.name}: {e}")
            continue
        
        # Frame variants
        for suffix in asset['frames']:
            frame_path = ASSETS_DIR / f"{asset['file']}_{suffix}.png"
            
            if not frame_path.exists():
                print(f"  ⚠️  Missing: {frame_path.name}")
                continue
            
            try:
                img = Image.open(frame_path)
                images_to_pack.append({
                    'key': f"{asset['key']}_{suffix}",
                    'path': frame_path,
                    'image': img,
                    'width': img.width,
                    'height': img.height
                })
            except Exception as e:
                print(f"  ❌ Error loading {frame_path.name}: {e}")
                continue
    
    if not images_to_pack:
        print(f"  ⚠️  No images to pack for {name}")
        return None, None
    
    # Sort by height (tallest first) for better packing efficiency
    images_to_pack.sort(key=lambda x: x['height'], reverse=True)
    
    # Pack images
    packer = ShelfPacker()
    packed = []
    
    for img_data in images_to_pack:
        pos = packer.pack(img_data['width'], img_data['height'])
        
        if pos is None:
            print(f"  ⚠️  Atlas full! {img_data['key']} doesn't fit (increase MAX_ATLAS_SIZE)")
            continue
        
        img_data['x'] = pos[0]
        img_data['y'] = pos[1]
        packed.append(img_data)
    
    # Get final atlas dimensions
    atlas_width, atlas_height = packer.get_dimensions()
    
    print(f"  📐 Atlas size: {atlas_width}×{atlas_height}px")
    print(f"  🎨 Packed {len(packed)} sprites")
    
    # Create atlas image
    atlas = Image.new('RGBA', (atlas_width, atlas_height), (0, 0, 0, 0))
    
    # Paste all sprites
    for img_data in packed:
        atlas.paste(img_data['image'], (img_data['x'], img_data['y']))
    
    # Generate Phaser JSON (texture atlas format)
    frames = {}
    
    for img_data in packed:
        frames[img_data['key']] = {
            'frame': {
                'x': img_data['x'],
                'y': img_data['y'],
                'w': img_data['width'],
                'h': img_data['height']
            },
            'rotated': False,
            'trimmed': False,
            'spriteSourceSize': {
                'x': 0,
                'y': 0,
                'w': img_data['width'],
                'h': img_data['height']
            },
            'sourceSize': {
                'w': img_data['width'],
                'h': img_data['height']
            }
        }
    
    atlas_json = {
        'frames': frames,
        'meta': {
            'app': 'Mimi vs. Math Atlas Packer',
            'version': '1.0',
            'image': f'{name}.png',
            'format': 'RGBA8888',
            'size': {'w': atlas_width, 'h': atlas_height},
            'scale': '1'
        }
    }
    
    return atlas, atlas_json


def main():
    print("╔═══════════════════════════════════════╗")
    print("║   Texture Atlas Packer                ║")
    print("╚═══════════════════════════════════════╝")
    
    # Create output directory
    ATLAS_OUTPUT_DIR.mkdir(exist_ok=True, parents=True)
    
    # Parse asset config
    print("\n📋 Parsing AssetConfig.js...")
    categories = parse_asset_config()
    
    for category, assets in categories.items():
        print(f"   {category}: {len(assets)} assets")
    
    # Create atlases
    atlases_created = 0
    
    for category, assets in categories.items():
        if not assets:
            continue
        
        atlas_name = f'atlas_{category}'
        atlas_img, atlas_json = create_atlas(atlas_name, assets)
        
        if atlas_img is None:
            continue
        
        # Save atlas PNG
        png_path = ATLAS_OUTPUT_DIR / f'{atlas_name}.png'
        atlas_img.save(png_path, 'PNG', optimize=True)
        print(f"  💾 Saved: {png_path.relative_to(REPO_ROOT)}")
        
        # Save atlas JSON
        json_path = ATLAS_OUTPUT_DIR / f'{atlas_name}.json'
        with open(json_path, 'w', encoding='utf-8') as f:
            json.dump(atlas_json, f, indent=2)
        print(f"  💾 Saved: {json_path.relative_to(REPO_ROOT)}")
        
        atlases_created += 1
    
    # Summary
    print("\n╔═══════════════════════════════════════╗")
    print("║   Packing Summary                     ║")
    print("╚═══════════════════════════════════════╝")
    print(f"  ✅ Created {atlases_created} texture atlases")
    print(f"  📁 Output: {ATLAS_OUTPUT_DIR.relative_to(REPO_ROOT)}")
    print()
    print("✨ Done! Texture atlases created.")
    print()
    print("Next steps:")
    print("  1. Update AssetConfig.js to set ATLAS_MODE = true")
    print("  2. Test in browser: python -m http.server 8080")
    
    return 0


if __name__ == '__main__':
    exit(main())

File: tools/pack_atlases.js
/**
 * pack_atlases.js — Pack PNG assets into texture atlases with Phaser JSON.
 * 
 * This script organizes all PNG sprites into multiple texture atlases:
 *   - atlas_characters.png/json — Mimi + enemies + NPCs
 *   - atlas_terrain.png/json    — floors, walls, decorations, landmarks
 *   - atlas_ui.png/json         — UI elements, hearts, items
 *   - atlas_backdrops.png/json  — battle backgrounds
 * 
 * Requires: npm install pngjs sharp
 * 
 * Usage:
 *     node tools/pack_atlases.js
 */

const fs = require('fs');
const path = require('path');
const { PNG } = require('pngjs');
const sharp = require('sharp');

const REPO_ROOT = path.join(__dirname, '..');
const ASSETS_DIR = path.join(REPO_ROOT, 'assets', 'sprites');
const ATLAS_OUTPUT_DIR = path.join(REPO_ROOT, 'assets', 'atlases');
const ASSET_CONFIG_PATH = path.join(REPO_ROOT, 'src', 'config', 'AssetConfig.js');

const MAX_ATLAS_SIZE = 2048;
const PADDING = 2;

/**
 * Simple shelf-based bin packing algorithm.
 */
class ShelfPacker {
    constructor(maxWidth = MAX_ATLAS_SIZE, maxHeight = MAX_ATLAS_SIZE) {
        this.maxWidth = maxWidth;
        this.maxHeight = maxHeight;
        this.shelves = [];
        this.currentShelf = null;
        this.usedHeight = 0;
    }
    
    pack(width, height) {
        const w = width + PADDING * 2;
        const h = height + PADDING * 2;
        
        // Try to fit on current shelf
        if (this.currentShelf) {
            const [shelfX, shelfY, shelfW, shelfH] = this.currentShelf;
            const nextX = shelfX + shelfW;
            
            if (nextX + w <= this.maxWidth && h <= shelfH) {
                this.currentShelf = [nextX, shelfY, shelfW + w, shelfH];
                return [nextX + PADDING, shelfY + PADDING];
            }
        }
        
        // Need a new shelf
        const newShelfY = this.usedHeight;
        
        if (newShelfY + h > this.maxHeight) {
            return null;  // Atlas full
        }
        
        this.currentShelf = [0, newShelfY, w, h];
        this.shelves.push(this.currentShelf);
        this.usedHeight = newShelfY + h;
        
        return [PADDING, newShelfY + PADDING];
    }
    
    getDimensions() {
        if (this.shelves.length === 0) {
            return [0, 0];
        }
        
        let maxWidth = Math.max(...this.shelves.map(s => s[0] + s[2]));
        const maxHeight = this.usedHeight;
        
        // Round up to power of 2
        let width = 1;
        while (width < maxWidth) width *= 2;
        
        let height = 1;
        while (height < maxHeight) height *= 2;
        
        return [
            Math.min(width, MAX_ATLAS_SIZE),
            Math.min(height, MAX_ATLAS_SIZE)
        ];
    }
}

/**
 * Parse AssetConfig.js to categorize assets.
 */
function parseAssetConfig() {
    const content = fs.readFileSync(ASSET_CONFIG_PATH, 'utf-8');
    
    const categories = {
        characters: [],
        terrain: [],
        ui: [],
        backdrops: []
    };
    
    const sectionMap = {
        'SPRITE_DEFS': 'characters',
        'UI_DEFS': 'ui',
        'TERRAIN_DEFS': 'terrain',
        'BACKDROP_DEFS': 'backdrops'
    };
    
    for (const [section, category] of Object.entries(sectionMap)) {
        const pattern = new RegExp(`export const ${section} = \\[(.*?)\\];`, 's');
        const match = content.match(pattern);
        
        if (!match) continue;
        
        const arrayContent = match[1];
        const objPattern = /\{[^}]+\}/g;
        const objects = arrayContent.match(objPattern) || [];
        
        for (const objStr of objects) {
            const keyMatch = objStr.match(/key:\s*'([^']+)'/);
            const fileMatch = objStr.match(/file:\s*'([^']+)'/);
            
            if (!keyMatch || !fileMatch) continue;
            
            const key = keyMatch[1];
            const file = fileMatch[1];
            
            const frames = [];
            const framesMatch = objStr.match(/frames:\s*\[([^\]]+)\]/);
            if (framesMatch) {
                const framesStr = framesMatch[1];
                const frameMatches = framesStr.matchAll(/'([^']+)'/g);
                for (const match of frameMatches) {
                    frames.push(match[1]);
                }
            }
            
            categories[category].push({ key, file, frames });
        }
    }
    
    return categories;
}

/**
 * Load a PNG image and return its data.
 */
async function loadPng(filePath) {
    const buffer = fs.readFileSync(filePath);
    return PNG.sync.read(buffer);
}

/**
 * Create a texture atlas from a list of assets.
 */
async function createAtlas(name, assets) {
    console.log(`\n📦 Creating ${name} atlas...`);
    
    // Collect all images to pack
    const imagesToPack = [];
    
    for (const asset of assets) {
        // Base image
        const pngPath = path.join(ASSETS_DIR, `${asset.file}.png`);
        
        if (!fs.existsSync(pngPath)) {
            console.log(`  ⚠️  Missing: ${path.basename(pngPath)}`);
            continue;
        }
        
        try {
            const img = await loadPng(pngPath);
            imagesToPack.push({
                key: asset.key,
                path: pngPath,
                image: img,
                width: img.width,
                height: img.height
            });
        } catch (e) {
            console.log(`  ❌ Error loading ${path.basename(pngPath)}: ${e.message}`);
            continue;
        }
        
        // Frame variants
        for (const suffix of asset.frames) {
            const framePath = path.join(ASSETS_DIR, `${asset.file}_${suffix}.png`);
            
            if (!fs.existsSync(framePath)) {
                console.log(`  ⚠️  Missing: ${path.basename(framePath)}`);
                continue;
            }
            
            try {
                const img = await loadPng(framePath);
                imagesToPack.push({
                    key: `${asset.key}_${suffix}`,
                    path: framePath,
                    image: img,
                    width: img.width,
                    height: img.height
                });
            } catch (e) {
                console.log(`  ❌ Error loading ${path.basename(framePath)}: ${e.message}`);
                continue;
            }
        }
    }
    
    if (imagesToPack.length === 0) {
        console.log(`  ⚠️  No images to pack for ${name}`);
        return null;
    }
    
    // Sort by height (tallest first)
    imagesToPack.sort((a, b) => b.height - a.height);
    
    // Pack images
    const packer = new ShelfPacker();
    const packed = [];
    
    for (const imgData of imagesToPack) {
        const pos = packer.pack(imgData.width, imgData.height);
        
        if (pos === null) {
            console.log(`  ⚠️  Atlas full! ${imgData.key} doesn't fit`);
            continue;
        }
        
        imgData.x = pos[0];
        imgData.y = pos[1];
        packed.push(imgData);
    }
    
    // Get final atlas dimensions
    const [atlasWidth, atlasHeight] = packer.getDimensions();
    
    console.log(`  📐 Atlas size: ${atlasWidth}×${atlasHeight}px`);
    console.log(`  🎨 Packed ${packed.length} sprites`);
    
    // Create atlas PNG
    const atlas = new PNG({ width: atlasWidth, height: atlasHeight });
    
    // Paste all sprites using pngjs
    for (const imgData of packed) {
        const img = imgData.image;
        for (let y = 0; y < img.height; y++) {
            for (let x = 0; x < img.width; x++) {
                const srcIdx = (img.width * y + x) << 2;
                const dstIdx = (atlasWidth * (imgData.y + y) + (imgData.x + x)) << 2;
                
                atlas.data[dstIdx] = img.data[srcIdx];         // R
                atlas.data[dstIdx + 1] = img.data[srcIdx + 1]; // G
                atlas.data[dstIdx + 2] = img.data[srcIdx + 2]; // B
                atlas.data[dstIdx + 3] = img.data[srcIdx + 3]; // A
            }
        }
    }
    
    // Generate Phaser JSON
    const frames = {};
    
    for (const imgData of packed) {
        frames[imgData.key] = {
            frame: {
                x: imgData.x,
                y: imgData.y,
                w: imgData.width,
                h: imgData.height
            },
            rotated: false,
            trimmed: false,
            spriteSourceSize: {
                x: 0,
                y: 0,
                w: imgData.width,
                h: imgData.height
            },
            sourceSize: {
                w: imgData.width,
                h: imgData.height
            }
        };
    }
    
    const atlasJson = {
        frames,
        meta: {
            app: 'Mimi vs. Math Atlas Packer',
            version: '1.0',
            image: `${name}.png`,
            format: 'RGBA8888',
            size: { w: atlasWidth, h: atlasHeight },
            scale: '1'
        }
    };
    
    return { atlas, atlasJson };
}

async function main() {
    console.log('╔═══════════════════════════════════════╗');
    console.log('║   Texture Atlas Packer                ║');
    console.log('╚═══════════════════════════════════════╝');
    
    // Create output directory
    if (!fs.existsSync(ATLAS_OUTPUT_DIR)) {
        fs.mkdirSync(ATLAS_OUTPUT_DIR, { recursive: true });
    }
    
    // Parse asset config
    console.log('\n📋 Parsing AssetConfig.js...');
    const categories = parseAssetConfig();
    
    for (const [category, assets] of Object.entries(categories)) {
        console.log(`   ${category}: ${assets.length} assets`);
    }
    
    // Create atlases
    let atlasesCreated = 0;
    
    for (const [category, assets] of Object.entries(categories)) {
        if (assets.length === 0) continue;
        
        // Skip backdrops - they're too large (800×600px each) to pack efficiently
        // Load them as individual PNGs instead
        if (category === 'backdrops') {
            console.log(`\n⏭️  Skipping ${category} (too large for atlas, will load individually)`);
            continue;
        }
        
        const atlasName = `atlas_${category}`;
        const result = await createAtlas(atlasName, assets);
        
        if (!result) continue;
        
        const { atlas, atlasJson } = result;
        
        // Save atlas PNG
        const pngPath = path.join(ATLAS_OUTPUT_DIR, `${atlasName}.png`);
        const buffer = PNG.sync.write(atlas);
        fs.writeFileSync(pngPath, buffer);
        console.log(`  💾 Saved: ${path.relative(REPO_ROOT, pngPath)}`);
        
        // Save atlas JSON
        const jsonPath = path.join(ATLAS_OUTPUT_DIR, `${atlasName}.json`);
        fs.writeFileSync(jsonPath, JSON.stringify(atlasJson, null, 2));
        console.log(`  💾 Saved: ${path.relative(REPO_ROOT, jsonPath)}`);
        
        atlasesCreated++;
    }
    
    // Summary
    console.log('\n╔═══════════════════════════════════════╗');
    console.log('║   Packing Summary                     ║');
    console.log('╚═══════════════════════════════════════╝');
    console.log(`  ✅ Created ${atlasesCreated} texture atlases`);
    console.log(`  📁 Output: ${path.relative(REPO_ROOT, ATLAS_OUTPUT_DIR)}`);
    console.log();
    console.log('✨ Done! Texture atlases created.');
    console.log();
    console.log('Next steps:');
    console.log('  1. Update AssetConfig.js to set ATLAS_MODE = true');
    console.log('  2. Test in browser: python -m http.server 8080');
    
    return 0;
}

main().then(code => process.exit(code));
